Document/0x05f-Testing-Local-Authentication.md

```diff
@@ -82,21 +84,66 @@ Make sure that the unlocked key is used during the application flow. For example
+Biometric authentication allows a convenient mechanism for authentication, but also introduces additional attack surface when using it. The Android developer documentation gives an interesting overview and indicators for [measuring biometric unlock security](https://source.android.com/security/biometric/measure#strong-weak-unlocks "Measuring Biometric Unlock Security").
+
+On the Android platform there are three different classes available for biometric authentication:
+
+- Android Q / 10 (API level 29) and higher: `BiometricManager`
+- Android P / 9 (API level 28) and higher: `BiometricPrompt`
+- Android 6.0 (API level 23) and higher: `FingerprintManager` (deprecated in Android P)
+
+<img src="Images/Chapters/0x05f/biometricprompt-architecture.png" width="500" alt="Biometric Auth in Android">
+
+The class [BiometricManager](https://developer.android.com/reference/kotlin/android/hardware/biometrics/BiometricManager "BiometricManager") can be used to verify if biometric hardware is available on the device and configured by the user. If that's the case, the class [BiometricPrompt] can be used to show a system-provided biometric dialog.
+
+The `BiometricPrompt` class is a significant improvement, as it allows to have a consistent UI for biometric authentication on Android and also supports more sensors than just fingerprint.
 
-Better security is achieved by using the fingerprint API in conjunction with the Android `KeyGenerator` class. With this method, a symmetric key is stored in the KeyStore and "unlocked" with the user's fingerprint. For example, to enable user access to a remote service, an AES key is created which encrypts the user PIN or authentication token. By calling `setUserAuthenticationRequired(true)` when creating the key, it is ensured that the user must re-authenticate to retrieve it. The encrypted authentication credentials can then be saved directly to regular storage on the the device (e.g. `SharedPreferences`). This design is a relatively safe way to ensure the user actually entered an authorized fingerprint. Note however that this setup requires the app to hold the symmetric key in memory during cryptographic operations, potentially exposing it to attackers that manage to access the app's memory during runtime.
+This is different to the `FingerprintManager` class that provided no UI and developers needed to build their own fingerprint UI and only fingerprint sensors are supported.

```

Comment:
Slightly re-phrased to improve the flow and implication regarding concepts UI/sensors.
```suggestion
This is different to the `FingerprintManager` class which only supports fingerprint sensors and provides no UI, forcing developers to build their own fingerprint UI.
```